datajoint/expression.py

```diff
@@ -348,11 +161,240 @@ def restrict(self, restriction):
         :param restriction: a sequence or an array (treated as OR list), another QueryExpression, an SQL condition
         string, or an AndList.
         """
-        assert is_true(restriction) or not self.heading.expressions or isinstance(self, GroupBy), \
-            "Cannot restrict a projection with renamed attributes in place."
-        self.restriction.append(restriction)
-        return self
+        attributes = set()
+        new_condition = make_condition(self, restriction, attributes)
+        if new_condition is True:
+            return self  # restriction has no effect, return the same object
+        # check that all attributes in condition are present in the query
+        try:
+            raise DataJointError("Attribute `%s` is not found in query." % next(
+                attr for attr in attributes if attr not in self.heading.names))
+        except StopIteration:
+            pass  # all ok
+        # If the new condition uses any new attributes, a subquery is required.
+        # However, Aggregation's HAVING statement works fine with aliased attributes.
+        need_subquery = isinstance(self, Union) or (
+                not isinstance(self, Aggregation) and self.heading.new_attributes)
+        if need_subquery:
+            result = self.make_subquery()
+        else:
+            result = copy.copy(self)
+            result._restriction = AndList(self.restriction)  # copy to preserve the original
+        result.restriction.append(new_condition)
+        result.restriction_attributes.update(attributes)
+        return result
+
+    def restrict_in_place(self, restriction):
+        self.__dict__.update(self.restrict(restriction).__dict__)
+
+    def __and__(self, restriction):
+        """
+        Restriction operator
+        :return: a restricted copy of the input argument
+        See QueryExpression.restrict for more detail.
+        """
+        return self.restrict(restriction)
+
+    def __xor__(self, restriction):
+        """
+        Restriction operator ignoring compatibility check.
+        """
+        if inspect.isclass(restriction) and issubclass(restriction, QueryExpression):
+            restriction = restriction()
+        if isinstance(restriction, Not):
+            return self.restrict(Not(PromiscuousOperand(restriction.restriction)))
+        return self.restrict(PromiscuousOperand(restriction))
+
+    def __sub__(self, restriction):
+        """
+        Inverted restriction
+        :return: a restricted copy of the input argument
+        See QueryExpression.restrict for more detail.
+        """
+        return self.restrict(Not(restriction))
 
+    def __neg__(self):
+        if isinstance(self, Not):
+            return self.restriction
+        return Not(self)
+
+    def __mul__(self, other):
+        """ join of query expressions `self` and `other` """
+        return self.join(other)
+
+    def __matmul__(self, other):
```

Comment:
This operator presents an issue on the MATLAB side since it is [not an allowed operator](https://www.mathworks.com/help/matlab/matlab_oop/implementing-operators-for-your-class.html) to overload. In the spirit of preserving consistency between clients, we should offer a means which feels similar when introducing a new operator. It doesn't appear as if we have a precedent (intentionally) for this.

Reply:
Each language will necessarily implement DataJoint differently. The Python implementation will use the best available operators. Matlab may find others.

Reply:
Ah yes; missed the implementation of `.join` with appropriate kwargs. That is the *actual* join operation with the `*` and `@` as appropriate aliases for the client language. In that case, we can mark this as resolved here, however, this will be problematic to address in MATLAB in a backward-compatible way... Proper implementation between the two DataJoint clients should have been:
- Join w/ semantic check: `*` for Python, `.*` for MATLAB
- Join w/o semantic check: `@` for Python, `*` for MATLAB

Perhaps the short-term solution would be that once we introduce the operation in MATLAB, we can simply choose not to select an operator alias but have the user structure queries with a specific order in mind.

This just bring further need to bring about `dj.expression(...)` so that we can explicitly apply a standard order to parsing operations so that user experiences a consistent way to generate queries between clients. Will file separate issues on these points.